src/geometry/manifold/bump_function.lean

```diff
@@ -0,0 +1,599 @@
+f.locally_finite.fintype_of_compact $ λ i, (f i).nonempty_support
+
+variables [t2_space M] [fintype f.ι]
+
+/-- Smooth embedding of `M` into `(E × ℝ) ^ f.ι`. -/
+def embedding_pi_tangent : C^∞⟮I, M; 𝓘(ℝ, f.ι → (E × ℝ)), f.ι → (E × ℝ)⟯ :=
+{ to_fun := λ x i, (f i x • ext_chart_at I (f.c i) x, f i x),
+  times_cont_mdiff_to_fun := times_cont_mdiff_pi_space.2 $ λ i,
+    ((f i).smooth_smul times_cont_mdiff_on_ext_chart_at).prod_mk_space ((f i).smooth) }
+
+local attribute [simp] lemma embedding_pi_tangent_coe :
+  ⇑f.embedding_pi_tangent = λ x i, (f i x • ext_chart_at I (f.c i) x, f i x) :=
+rfl
+
+lemma embedding_pi_tangent_inj_on : inj_on f.embedding_pi_tangent s :=
+begin
+  intros x hx y hy h,
+  simp only [embedding_pi_tangent_coe, funext_iff] at h,
+  obtain ⟨h₁, h₂⟩ := prod.mk.inj_iff.1 (h (f.ind x hx)),
+  rw [f.apply_ind x hx] at h₂,
+  rw [← h₂, f.apply_ind x hx, one_smul, one_smul] at h₁,
+  have := f.mem_ext_chart_at_source_of_eq_one h₂.symm,
+  exact (ext_chart_at I (f.c _)).inj_on (f.mem_ext_chart_at_ind_source x hx) this h₁
+end
+
+lemma embedding_pi_tangent_injective (f : smooth_bump_covering I (univ : set M))
+  [fintype f.ι] :
+  injective f.embedding_pi_tangent :=
+injective_iff_inj_on_univ.2 f.embedding_pi_tangent_inj_on
+
+lemma comp_embedding_pi_tangent_mfderiv (x : M) (hx : x ∈ s) :
+  ((continuous_linear_map.fst ℝ E ℝ).comp
+    (@continuous_linear_map.proj ℝ _ f.ι (λ _, E × ℝ) _ _ (λ _, infer_instance) (f.ind x hx))).comp
+      (mfderiv I 𝓘(ℝ, f.ι → (E × ℝ)) f.embedding_pi_tangent x) =
+  mfderiv I I (chart_at H (f.c (f.ind x hx))) x :=
+begin
+  set L := ((continuous_linear_map.fst ℝ E ℝ).comp
+    (@continuous_linear_map.proj ℝ _ f.ι (λ _, E × ℝ) _ _ (λ _, infer_instance) (f.ind x hx))),
+  have := (L.has_mfderiv_at.comp x (f.embedding_pi_tangent.mdifferentiable_at.has_mfderiv_at)),
+  convert has_mfderiv_at_unique this _,
+  refine (has_mfderiv_at_ext_chart_at I (f.mem_chart_at_ind_source x hx)).congr_of_eventually_eq _,
+  refine (f.eventually_eq_one x hx).mono (λ y hy, _),
+  simp only [embedding_pi_tangent_coe, continuous_linear_map.coe_comp', (∘),
+    continuous_linear_map.coe_fst', continuous_linear_map.proj_apply],
+  rw [hy, pi.one_apply, one_smul]
+end
+
+lemma embedding_pi_tangent_ker_mfderiv (x : M) (hx : x ∈ s) :
+  (mfderiv I 𝓘(ℝ, f.ι → (E × ℝ)) f.embedding_pi_tangent x).ker = ⊥ :=
+begin
+  apply bot_unique,
+  rw [← (mdifferentiable_chart I (f.c (f.ind x hx))).ker_mfderiv_eq_bot
+    (f.mem_chart_at_ind_source x hx), ← comp_embedding_pi_tangent_mfderiv],
+  exact linear_map.ker_le_ker_comp _ _
+end
+
+lemma embedding_pi_tangent_injective_mfderiv (x : M) (hx : x ∈ s) :
+  injective (mfderiv I 𝓘(ℝ, f.ι → (E × ℝ)) f.embedding_pi_tangent x) :=
+linear_map.ker_eq_bot.1 (f.embedding_pi_tangent_ker_mfderiv x hx)
+
+end embedding
+
+/-- Baby version of the Whitney weak embedding theorem: if `M` admits a finite covering by
+supports of bump functions, then for some `n` it can be embedded into the `n`-dimensional
```

Comment:
You are not proving that you have an embedding, only an injective immersion. When the space is compact, this can automatically be upgraded to an embedding, but not in general (think of the real line that you immerse in R^2 by folding one end to come back to a point in the image)

Reply:
I'll upgrade it to an embedding using `continuous.closed_embedding`.